src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt

```diff
@@ -3,77 +3,73 @@ package it.unibo.protelis.web.execution.simulated
 
-  init {
-    logger.debug("Simulation container created")
-  }
-
-  constructor(sourceCode: String, monitor: ProtelisObserver) : this() {
-    GlobalScope.launch { setupAwait(sourceCode, monitor) }.asCompletableFuture().join()
-  }
+  override fun setup(sourceCode: String, monitor: ProtelisObserver) {
+    if (alchemistEngine != null) {
+      stop()
+    }
 
-  override suspend fun setupAwait(sourceCode: String, monitor: ProtelisObserver) {
-    if (alchemistEngine.get() != null) stopAwait()
-    val sim = setupSimulationAsync().await()
-    alchemistEngine.set(sim)
+    alchemistEngine = setupSimulation()
     setCode(sourceCode)
-    alchemistEngine.get()
+    alchemistEngine
       ?.addOutputMonitor(ProtelisUpdateAdapter(monitor))
       ?: throw IllegalStateException("Simulation was not set up correctly")
-    alchemistEngine.get()
-      ?.awaitFor(Status.INIT)
+    alchemistEngine
+      ?.waitForAndCheck(Status.INIT)
       ?: throw IllegalStateException("Simulation was not set up correctly")
-    GlobalScope.launch { sim.run() }
+    GlobalScope.launch() { alchemistEngine?.run() ?: throw IllegalStateException() }
     logger.debug("Simulation Engine set up correctly")
   }
 
-  override suspend fun startAwait() {
-    checkNotNull(alchemistEngine.get())
-    alchemistEngine.get()?.play()
-    alchemistEngine.get()?.awaitFor(Status.RUNNING) // TODO: should I wait RUNNING instead?
+  override fun start() {
+    checkNotNull(alchemistEngine)
+    alchemistEngine
+      ?.play()
+      ?: throw IllegalStateException()
+    alchemistEngine
+      ?.waitForAndCheck(Status.RUNNING)
+      ?: throw IllegalStateException()
     logger.debug("Simulation started")
   }
 
-  override suspend fun stopAwait() {
-    checkNotNull(alchemistEngine.get())
-    alchemistEngine.get()?.terminate()
-    alchemistEngine.get()?.awaitFor(Status.TERMINATED)
-    alchemistEngine.set(null)
+  override fun stop() {
+    checkNotNull(alchemistEngine)
+    alchemistEngine
+      ?.terminate()
+      ?: throw IllegalStateException()
+    alchemistEngine
+      ?.waitForAndCheck(Status.TERMINATED)
+      ?: throw IllegalStateException()
+    alchemistEngine = null
     logger.debug("Simulation stopped and deleted")
   }
 
   /** Asynchronously build a new simulation from a loader of YAML files. */
-  private fun setupSimulationAsync(): Deferred<Engine<Any /* TODO */, Euclidean2DPosition>> = GlobalScope.async {
-    val loader = YamlLoader(this.javaClass.classLoader.getResourceAsStream("simulation.yml"))
-    loader.getDefault<Any /* TODO */, Euclidean2DPosition>()
-    Engine<Any /* TODO */, Euclidean2DPosition>(
-      loader.getDefault(),
+  private fun setupSimulation(): Engine<Any, Euclidean2DPosition> =
+    Engine<Any, Euclidean2DPosition>(
+      YamlLoader(this.javaClass.classLoader.getResourceAsStream("simulation.yml")).getDefault(),
       Long.MAX_VALUE,
       DoubleTime(Double.POSITIVE_INFINITY)
     )
-  }
 
+  /** Update code in nodes. */
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [CommentOverPrivateFunction - [setCode] at /src/src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt:70:3](https://app.codacy.com/manual/NiccoMlt/Protelis-Web/pullRequest?prid=5040950)